File: python/owrt_monitor/config_diff.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

# Sentinel — distinguishes "key present with value None" from "key absent".
_MISSING = object()


@dataclass(frozen=True)
class ConfigChange:
    """One difference between two redacted config snapshots.

    `path` is dotted (e.g. `tests.smoke[2].command`), `old` is the previous
    value, `new` is the new value, with `<missing>` placeholders when a side
    is absent (added or removed key).
    """

    path: str
    old: Any
    new: Any


def diff_configs(old: Any, new: Any, *, prefix: str = "") -> list[ConfigChange]:
    """Recursively diff two redacted config snapshots.

    Rules:
      - Dicts: recurse into common keys; keys on only one side become
        added/removed entries.
      - Lists: when lengths differ, record the whole list as one change.
        When equal, recurse element-wise (path uses `[i]` suffix).
      - Scalars: equal → skip, differ → one change.

    The output is sorted by path for stable display.
    """
    changes = _walk(old, new, prefix)
    changes.sort(key=lambda c: c.path)
    return changes


def _walk(old: Any, new: Any, prefix: str) -> list[ConfigChange]:
    if isinstance(old, dict) and isinstance(new, dict):
        out: list[ConfigChange] = []
        for key in set(old) | set(new):
            sub_path = f"{prefix}.{key}" if prefix else key
            sub_old = old.get(key, _MISSING)
            sub_new = new.get(key, _MISSING)
            if sub_old is _MISSING:
                out.append(ConfigChange(sub_path, "<missing>", sub_new))
            elif sub_new is _MISSING:
                out.append(ConfigChange(sub_path, sub_old, "<missing>"))
            else:
                out.extend(_walk(sub_old, sub_new, sub_path))
        return out
    if isinstance(old, list) and isinstance(new, list):
        if len(old) != len(new):
            return [ConfigChange(prefix or "<root>", old, new)]
        out = []
        for index, (a, b) in enumerate(zip(old, new, strict=True)):
            out.extend(_walk(a, b, f"{prefix}[{index}]"))
        return out
    if old != new:
        return [ConfigChange(prefix or "<root>", old, new)]
    return []


@dataclass(frozen=True)
class ConfigDiffSummary:
    """Compact summary suitable for event payloads / report rendering."""

    total: int
    sample: list[ConfigChange] = field(default_factory=list)


def summarize(changes: list[ConfigChange], *, sample_limit: int = 20) -> ConfigDiffSummary:
    return ConfigDiffSummary(
        total=len(changes),
        sample=list(changes[:sample_limit]),
    )

File: internal/configdiff/configdiff.go
// Package configdiff is the Go port of python/owrt_monitor/config_diff.py: it
// recursively diffs two redacted config snapshots into a path-sorted list of
// changes, used to surface what changed since the last successful run.
package configdiff

import (
	"fmt"
	"reflect"
	"sort"
)

// Missing is the placeholder shown when a key exists on only one side.
const Missing = "<missing>"

// Change is one difference between two snapshots. Path is dotted, e.g.
// "tests.smoke[2].command".
type Change struct {
	Path string `json:"path"`
	Old  any    `json:"old"`
	New  any    `json:"new"`
}

// Diff recursively diffs old vs new and returns changes sorted by path
// (config_diff.diff_configs).
func Diff(old, new any) []Change {
	changes := walk(old, new, "")
	sort.Slice(changes, func(i, j int) bool { return changes[i].Path < changes[j].Path })
	return changes
}

func walk(old, new any, prefix string) []Change {
	oldMap, oldIsMap := old.(map[string]any)
	newMap, newIsMap := new.(map[string]any)
	if oldIsMap && newIsMap {
		var out []Change
		for _, key := range unionKeys(oldMap, newMap) {
			subPath := key
			if prefix != "" {
				subPath = prefix + "." + key
			}
			subOld, hasOld := oldMap[key]
			subNew, hasNew := newMap[key]
			switch {
			case !hasOld:
				out = append(out, Change{subPath, Missing, subNew})
			case !hasNew:
				out = append(out, Change{subPath, subOld, Missing})
			default:
				out = append(out, walk(subOld, subNew, subPath)...)
			}
		}
		return out
	}

	oldList, oldIsList := old.([]any)
	newList, newIsList := new.([]any)
	if oldIsList && newIsList {
		if len(oldList) != len(newList) {
			return []Change{{rootOr(prefix), old, new}}
		}
		var out []Change
		for i := range oldList {
			out = append(out, walk(oldList[i], newList[i], fmt.Sprintf("%s[%d]", prefix, i))...)
		}
		return out
	}

	if !reflect.DeepEqual(old, new) {
		return []Change{{rootOr(prefix), old, new}}
	}
	return nil
}

// Summary is a compact view suitable for event payloads (config_diff.summarize).
type Summary struct {
	Total  int      `json:"total"`
	Sample []Change `json:"sample"`
}

// Summarize returns the total change count plus a bounded sample.
func Summarize(changes []Change, sampleLimit int) Summary {
	if sampleLimit < 0 {
		sampleLimit = 0
	}
	sample := changes
	if len(sample) > sampleLimit {
		sample = sample[:sampleLimit]
	}
	if sample == nil {
		sample = []Change{}
	}
	return Summary{Total: len(changes), Sample: sample}
}

func unionKeys(a, b map[string]any) []string {
	seen := map[string]bool{}
	var keys []string
	for k := range a {
		if !seen[k] {
			seen[k] = true
			keys = append(keys, k)
		}
	}
	for k := range b {
		if !seen[k] {
			seen[k] = true
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)
	return keys
}

func rootOr(prefix string) string {
	if prefix == "" {
		return "<root>"
	}
	return prefix
}
